### PR TITLE
Add support for array type completions #504

### PR DESCRIPTION
**Pull request by gayanper**
Enhance the existing support for discovering array component type by
adding extra information into CompletionProposal so that the consumers
can decide how the edits should be handled in the editors like jdt.ui
and jdt.ls

### CODE REVIEW

**Comment by gayanper** [comment]
@rgrunber @stephan-herrmann I add both of you from jdt and jdt.ls perspective. My plan is to initially add editor support in jdt.ls for array completions.

**Comment by jdneo** [comment]
Vote for this one. I had a [try](https://git.eclipse.org/r/c/jdt/eclipse.jdt.core/+/191208) to support array completion

This one look neater and does not require to create a new proposal type 👍

**Comment by stephan-herrmann** [comment]
@gayanper just a quick ACK: saw your request and will try to find time to look into it soonish.

File: org.eclipse.jdt.core/codeassist/org/eclipse/jdt/internal/codeassist/CompletionEngine.java [hunk 1]
```diff
@@ -12083,135 +12083,157 @@ private void findTypesFromExpectedTypes(char[] token, Scope scope, ObjectVector
...
+				boolean receiverIsArrayType = false;
 				if(this.expectedTypes[i] instanceof ReferenceBinding) {
-					ReferenceBinding refBinding = (ReferenceBinding)this.expectedTypes[i];
+					refBinding = (ReferenceBinding) this.expectedTypes[i];
+				} else if (this.expectedTypes[i] instanceof ArrayBinding) {
+					receiverIsArrayType = true;
+					TypeBinding leaf = ((ArrayBinding) this.expectedTypes[i]).leafComponentType();
+					refBinding = this.lookupEnvironment
+							.getType(CharOperation.splitOn('.', CharOperation.concat(leaf.qualifiedPackageName(),
+									leaf.qualifiedSourceName(), '.')));
```
Comment (stephan-herrmann) [inline org.eclipse.jdt.core/codeassist/org/eclipse/jdt/internal/codeassist/CompletionEngine.java:12095]: I'm not totally sure what's going on here:
* why the detour from an existing binding via its name to a new binding?
* what if `refBinding` is not a valid type binding afterwards?
* I'm missing an explicit check for array-of-primitive, perhaps that's what the detour achieves in an obscure way? :)

**Review by stephan-herrmann — COMMENTED** [review]
Sorry for the delay.

The change generally looks good. See one inline question.

Confidence could perhaps still be increased by some new tests, and/or a companion PR to JDT/UI where the new flag is evaluated to bring some added value.

File: org.eclipse.jdt.core/codeassist/org/eclipse/jdt/internal/codeassist/CompletionEngine.java (hunk 1 above)
Comment (gayanper) [reply org.eclipse.jdt.core/codeassist/org/eclipse/jdt/internal/codeassist/CompletionEngine.java:12095]: I will check this again @stephan-herrmann and update. I think I have to refresh my approach here :).

**Comment by gayanper** [comment]
@stephan-herrmann I reworked the solution, eliminating some white space changes and improving on the logic as you have pointed out in earlier fix.

**Comment by stephan-herrmann** [comment]
> [gayanper](https://github.com/gayanper) [force-pushed](https://github.com/eclipse-jdt/eclipse.jdt.core/compare/f40e24d63d1a06b26d6240a41ac0278ed8a4443e..a89797dd95c78159de2bdc98372e1fe19e2ed6cc) the add_array_completions branch from[ f40e24d ](https://github.com/eclipse-jdt/eclipse.jdt.core/commit/f40e24d63d1a06b26d6240a41ac0278ed8a4443e)to a89797d yesterday

Didn't we agree on avoiding force push to facilitate re-reviews? :smile:

Anyway, I'm looking at this now ...

**Review by stephan-herrmann — APPROVED** [review]
From jdt.core side this looks good, now.

I recall in my previous review I was asking about seeing this in action in a client (just the review appears to be gone due to force push).

Anyway, I still suggest to demonstrate in either jdt.ls or jdt.ui that the new API serves the purpose. Merging this PR early in 4.28 should allow as to make any required changes before the next release...

**Comment by gayanper** [comment]
> > [gayanper](https://github.com/gayanper) [force-pushed](https://github.com/eclipse-jdt/eclipse.jdt.core/compare/f40e24d63d1a06b26d6240a41ac0278ed8a4443e..a89797dd95c78159de2bdc98372e1fe19e2ed6cc) the add_array_completions branch from[ f40e24d ](https://github.com/eclipse-jdt/eclipse.jdt.core/commit/f40e24d63d1a06b26d6240a41ac0278ed8a4443e)to [a89797d](https://github.com/eclipse-jdt/eclipse.jdt.core/commit/a89797dd95c78159de2bdc98372e1fe19e2ed6cc) yesterday
> 
> Didn't we agree on avoiding force push to facilitate re-reviews? 😄
> 
> Anyway, I'm looking at this now ...

Yes we agreed, But for this change since I reworked to avoid unwanted whitespace changes it was easy for me to start from a new branch and push to the same PR remote branch to keep the PR open. Didn't want to spend lot of time fiddling with git. Sorry for the inconvenience with force push.

Your previous commands are there, they are just collapsed since I resolved the discussion thread and they are also outdated since the code is changed.

**Comment by gayanper** [comment]
> From jdt.core side this looks good, now.
> 
> I recall in my previous review I was asking about seeing this in action in a client (just the review appears to be gone due to force push).
> 
> Anyway, I still suggest to demonstrate in either jdt.ls or jdt.ui that the new API serves the purpose. Merging this PR early in 4.28 should allow as to make any required changes before the next release...

Since I now use vscode as my main driver I can work on adding the support into jdt.ls. But with this change for unsupported consumer of JDT there are not significant changes. Here is a screencast

https://user-images.githubusercontent.com/7471994/222981296-0d0762a2-ff5d-473d-bdde-3a426e3d5ba3.mp4

as you can see it suggest the type reference only and for primitives, I think we need to handle that from jdt.ui side.

**Pull request merged**